04-onion-routing.md

```diff
@@ -488,24 +560,41 @@ The node returning the message builds a return packet consisting of the followin
    * [`2`:`pad_len`]
    * [`pad_len`:`pad`]
 
-Where `hmac` is an HMAC authenticating the remainder of the packet, with a key using the above key generation with key type `um`, `failuremsg` is defined below, and `pad` as extra bytes to conceal length.
+Where `hmac` is an HMAC authenticating the remainder of the packet, with a key
+generated using the above process, with key type `um`, `failuremsg` as defined
+below, and `pad` as the extra bytes used to conceal length.
 
-The node SHOULD set `pad` such that the `failure_len` plus `pad_len` is equal to 256.
-This is 118 bytes longer than then the longest currently-defined message.
+The node:
+  - SHOULD set `pad` such that the `failure_len` plus `pad_len` is equal to 256.
+    - Note: this value is 118 bytes longer than the longest currently-defined
+    message.
 
 The node then generates a new key, using the key type `ammag`.
-This key is then used to generate a pseudo-random stream, which is then applied to the packet using `XOR`.
+This key is then used to generate a pseudo-random stream, which is then applied
+to the packet using `XOR`.
 
 The obfuscation step is repeated by every node on the return path.
-Upon receiving a packet the node will generate its `ammag`, generate the pseudo-random byte stream and apply it to the packet before forwarding.
-
-The origin node detects that it is the final hop of the return message since it was the origin of the corresponding forward packet.
-When an origin node receive an error message matching a transfer it initiated, i.e. it cannot forward the error any further, the node generates the `ammag` and `um` keys for each hop in the route.
-The node then iteratively decrypts the error message with each of the `ammag` keys and computes the HMAC with the `um` keys.
-The origin node can detect the sender of the error message by matching the `hmac` field with the computed HMAC.
-Once the original message has been decrypted it SHOULD be stored in a copy and the node SHOULD continue decrypting, until the loop has been repeated 20 times, using a constant `ammag` and `um` keys to obfuscate the route length.
-
-The association between forward and return packet is handled outside of the protocol, e.g. by association to an HTLC in a payment channel.
+Upon receiving a packet, the node will generate its `ammag`, generate the
+pseudo-random byte stream, and apply it to the packet before forwarding.
+
+The origin node detects that it is the final hop of the return message, since it
+was the origin of the corresponding forward packet.
+When an origin node receive an error message matching a transfer it initiated,
```

Comment:
The origin node is able to detect that it's the intended final recipient of the
return message, because of course, it was the originator of the corresponding
forward packet.